ipfs/provider.go

```diff
@@ -3,8 +3,40 @@ package ipfs
 import (
 	"io"
 
-	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	"github.com/ipfs/go-ipfs/core"
+	ipld "github.com/ipfs/go-ipld-format"
+	"github.com/libp2p/go-libp2p-core/routing"
 )
 
 // APIProvider allows customizable IPFS core APIs.
-type APIProvider func() (coreiface.APIDagService, io.Closer, error)
+type APIProvider interface {
+	DAG() ipld.DAGService
+	Blockstore() blockstore.Blockstore
+	Routing() routing.ContentRouting
+
+	io.Closer
+}
```

Comment:
Let's just make APIProvider a NodeProvider that just returns struct, this way nothing has to be changed at all.

Reply:
but then we have use a full node to easily mock, and that causes issues in the CI

Reply:
* Current code in master uses mock node and no we have no issues
* Even if we had them, Ci issues should be resolved by giving CI better resources

Reply:
> Current code in master uses mock node and no we have no issues

this is true for our `ipld` package, but not true for everything else that use the dag only mock in `ipfs.Mock`. Using a struct would prevent us from being able to pass in a mock. This is actually the reason why I liked it so much when you introduced `APIProvider`, and I still think it's a really good idea. That and it let's us expose only the portions of the ipfs API that we need.

#406 tests what happens to the CI when running a fully mocked node.

> Even if we had them, Ci issues should be resolved by giving CI better resources

I still think that using the APIProvider is the solution we should go with even when we do add more resources, if only because it gives us control over what is and isn't exposed.